File: app/src/main/java/com/silverpants/movieapp/datafactory/MovieDataSource.java
package com.silverpants.movieapp.datafactory;


import com.silverpants.movieapp.Keys;
import com.silverpants.movieapp.pojo.Discover;
import com.silverpants.movieapp.pojo.Result;
import com.silverpants.movieapp.retrofit.RetroFitRepository;
import com.silverpants.movieapp.retrofit.RetroFitApi;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.paging.PageKeyedDataSource;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MovieDataSource extends PageKeyedDataSource<Integer, Result> {

    private int year;
    private String sort_by;
    private RetroFitApi service;

    public MovieDataSource() {
        this(2019, Keys.SORT_BY_VALUES[0]);
    }

    public MovieDataSource(int year, String sort_by) {
        this.year = year;
        this.sort_by = sort_by;
        service = RetroFitRepository.getInstance().getRetroFitApi();
    }


    @Override
    public void loadInitial(@NonNull LoadInitialParams<Integer> params, @NonNull LoadInitialCallback<Integer, Result> callback) {
        Call<Discover> discover;
        discover = service.getDiscover(1, year, sort_by);
        discover.enqueue(new Callback<Discover>() {
            @Override
            public void onResponse(Call<Discover> call, Response<Discover> response) {
                List<Result> ls = response.body().getResults();
                callback.onResult(ls, null, 2);
            }

            @Override
            public void onFailure(Call<Discover> call, Throwable t) {
            }
        });


    }

    @Override
    public void loadBefore(@NonNull LoadParams<Integer> params, @NonNull LoadCallback<Integer, Result> callback) {
        //not required

    }

    @Override
    public void loadAfter(@NonNull LoadParams<Integer> params, @NonNull LoadCallback<Integer, Result> callback) {
        Call<Discover> discover;
        discover = service.getDiscover(params.key, year, sort_by);
        discover.enqueue(new Callback<Discover>() {
            @Override
            public void onResponse(Call<Discover> call, Response<Discover> response) {
                List<Result> ls = response.body().getResults();
                callback.onResult(ls, params.key + 1);
            }

            @Override
            public void onFailure(Call<Discover> call, Throwable t) {

            }
        });

    }
}

File: app/src/main/java/com/silverpants/movieapp/datafactory/MovieDataSource.kt
package com.silverpants.movieapp.datafactory

import com.silverpants.movieapp.pojo.discover.Discover
import com.silverpants.movieapp.pojo.discover.Result
import com.silverpants.movieapp.retrofit.RetroFitRepository
import com.silverpants.movieapp.retrofit.RetroFitApi
import androidx.paging.PageKeyedDataSource
import com.silverpants.movieapp.SORT_BY_VALUES
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MovieDataSource(var year: Int = 2019,var sort_by: String = SORT_BY_VALUES[0]) : PageKeyedDataSource<Int, Result>() {
    private val service: RetroFitApi = RetroFitRepository.retroFitApi


    override fun loadInitial(params: LoadInitialParams<Int>, callback: LoadInitialCallback<Int, Result>) {
        val discover: Call<Discover> = service.getDiscover(1, year, sort_by)
        discover.enqueue(object : Callback<Discover> {
            override fun onResponse(call: Call<Discover>, response: Response<Discover>) {
                val ls = response.body()!!.results
                callback.onResult(ls!!, null, 2)
            }

            override fun onFailure(call: Call<Discover>, t: Throwable) {}
        })


    }

    override fun loadBefore(params: LoadParams<Int>, callback: LoadCallback<Int, Result>) {
        //not required

    }

    override fun loadAfter(params: LoadParams<Int>, callback: LoadCallback<Int, Result>) {
        val discover: Call<Discover> = service.getDiscover(params.key, year, sort_by)
        discover.enqueue(object : Callback<Discover> {
            override fun onResponse(call: Call<Discover>, response: Response<Discover>) {
                val ls = response.body()!!.results
                callback.onResult(ls!!, params.key + 1)
            }

            override fun onFailure(call: Call<Discover>, t: Throwable) {

            }
        })

    }
}
